manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,857 @@
+	// will cause us to enter a bad state.
+
+	for _, config := range local.nw.IPAM.Configs {
+		// these things can return errors, but we literally can't do anything
+		// about it if they do, so just ignore it.
+		ipam.ReleaseAddress(local.pools[config.Subnet], net.ParseIP(config.Gateway))
+		ipam.ReleasePool(local.pools[config.Subnet])
+	}
+	// remove the IPAM config. this may be useful for "rolling back" a network
+	// allocation.
+	// NOTE(dperny): doing this can cause a data race. i believe it may be
+	// possible that the same network object is shared across event receivers.
+	// network.IPAM = nil
+	delete(a.networks, network.ID)
+}
+
+// AllocateVIPs allocates the VIPs for the provided endpoint and network ids.
+// The endpoint is assumed to be in resolution mode
+// api.ResolutionModeVirtualIP, which should be verified before calling. This
+// is the case because the endpoint object may reflect an older spec, instead
+// of the current one.
+//
+// If successful, this method will mutate the provided endpoint, but will not
+// otherwise do so.
+func (a *allocator) AllocateVIPs(endpoint *api.Endpoint, networkIDs map[string]struct{}) (rerr error) {
+	// first, go through and check that every network we want a VIP for is
+	// allocated. if not, return an error. We can't allocate VIPs until the
+	// network is allocated
+	for nwid := range networkIDs {
+		if _, ok := a.networks[nwid]; !ok {
+			return errors.ErrDependencyNotAllocated("network", nwid)
+		}
+	}
+
+	// now compute the changes
+	allocate := []string{}
+	// keep is the set of all virtual IPs we'll retain between the previous and
+	// current spec. we make it with a capacity the same as the current vips
+	// because typically, the vips in use won't change much, and we can avoid
+	// allocation by using this as a guess
+	keep := make([]*api.Endpoint_VirtualIP, 0, len(endpoint.VirtualIPs))
+	deallocate := []*api.Endpoint_VirtualIP{}
+	// first, we need to figure out if any virtual IPs are being removed
+	// continues are bad and hard to follow, so here's the plain english
+	// explanation:
+	// for every VIP currently allocated
+	//     go through the list of desired network IDs
+	//         if a network ID matches the ID on the VIP
+	//             then we're keeping this vip, so add it to the keep list and
+	//             go to the next VIPs, skipping the bottom of this loop
+	//     if we get to this point, then we have been through every desired
+	//     network ID and not found one matching the one on this VIP, so we
+	//     can add it to the list of VIPs to deallocate
+	for _, vip := range endpoint.VirtualIPs {
+		if _, ok := networkIDs[vip.NetworkID]; ok {
+			keep = append(keep, vip)
+		} else {
+			deallocate = append(deallocate, vip)
+		}
+	}
+
+	// now figure out which new network IDs we've added, which is the same loop
+	// above but swapped around to check network IDs against VIPs
+newvips:
+	for nwid := range networkIDs {
```

Comment:
No, because this vip loop is part of a quadratic check for which networks have vips already allocated and which do not. We can probably optimize this by combining this logic with the allocation logic, but I think that optimization would be out of scope for this revision.

Reply:
Right, what I mean is that this inner loop is going through every VIP to find a network ID that matches the network ID in the outer loop.

However, the loop above is also looping through every VIP, to see which ones to keep and deallocate.  In the upper loop, if we build up a mapping of network IDs to `struct{}{}`, we'll have a set of which network IDs are specified by the VIPS that we can use to look up here, instead of having an inner loop, if that makes sense.

e.g:

```golang
        nwidsInVIPS := make(map[string]struct{}{})
	for _, vip := range endpoint.VirtualIPs {
		if _, ok := networkIDs[vip.NetworkID]; ok {
			keep = append(keep, vip)
		} else {
			deallocate = append(deallocate, vip)
		}
		nwidsInVIPS[vip.NetworkID] = struct{}{}
	}

	for nwid := range networkIDs {
		if _, ok := nwidsInVIPS[vip.NetworkID]; !ok {
  		    allocate = append(allocate, nwid)
  		}
	}
```

Reply:
ahhhh... i see. let me put this in and run it through the wash.

Reply:
Done this.